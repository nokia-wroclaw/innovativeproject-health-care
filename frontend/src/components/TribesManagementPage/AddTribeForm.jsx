import React, { useState } from "react";
import { connect } from "react-redux";
import { Formik } from "formik";
import { Button, Form } from "semantic-ui-react";
import { addTribe } from "../../store/actions/tribes";

export const AddTribeForm = props => {
  const [loading, setLoading] = useState(false);

  return (
    <Formik
      initialValues={{ tribeName: "" }}
      validate={values => {
        let errors = {};
        if (!values.tribeName) errors.tribeName = "Tribe name required";
        return errors;
      }}
      onSubmit={(values, { setSubmitting }) => {
        setLoading(true);
        props.addTribe(values.tribeName).then(() => {
          setLoading(false);
          setSubmitting(false);
        });
      }}
    >
      {({ errors, handleChange, handleSubmit, isSubmiting }) => (
        <Form onSubmit={handleSubmit}>
          <Form.Field>
            <Form.Input
              onChange={handleChange}
              name="tribeName"
              placeholder="Tribe name"
              disabled={isSubmiting}
              error={!!errors.tribeName}
            />
          </Form.Field>

          <Button
            type="submit"
            floated="right"
            color="green"
            disabled={isSubmiting}
            loading={loading}
          >
            Add
          </Button>
        </Form>
      )}
    </Formik>
  );
};

export default connect(
  null,
  { addTribe }
)(AddTribeForm);
