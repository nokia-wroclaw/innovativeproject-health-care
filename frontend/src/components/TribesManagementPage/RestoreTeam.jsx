import React, { useState, useCallback } from "react";
import { connect } from "react-redux";
import { Modal, Button, Container, Segment, Message, SegmentGroup } from "semantic-ui-react";
import {
    addTeamToTribe,
  } from "../../store/actions/tribes";
import {
  restoreTeamToTribe
  } from "../../store/actions/teams";
import "../../styles/common.css";

const RestoreTeam = ({ isOpen, tribe, teamName, close, ...props }) => {
    const deletedTeams = tribe.teams.filter(team => team.deleted === true);
    const [selectedTeam, selectTeam] = useState(null);
    const [hovered, setHovered] = useState(null);

    const handleAddNewTeam = async () => {
      await props.addTeamToTribe(tribe, teamName);
      close();
    }

    const handleRestoreTeam = async (team) => {
      if (selectedTeam) {
        await props.restoreTeamToTribe(team);
        close();
      }
    }

    const handleSelectTeam = team => 
      selectTeam(team);

    const handleSetHovered = team => 
      setHovered(team);

    return(
      <Modal open={isOpen} onClose={close}>
        <Modal.Content>
          <Message>
            <Message.Header>
            Team with this name was created in the past. You can restore any team from the list or create a new one.
            </Message.Header>
          </Message>
          <Segment.Group>
          {deletedTeams
            ? deletedTeams.filter(team => team.name === teamName).map(team => (
              <SegmentGroup 
                horizontal
                key={team.id}
                onClick={() => handleSelectTeam(team)} 
                onMouseEnter={() => handleSetHovered(team)} 
                onMouseLeave={() => handleSetHovered(null)}
                className={(hovered === team ? " hovered" : "") + (selectedTeam === team ? " clicked" : "")}
              >
                <Segment>{team.name}</Segment>
                <Segment>{team.deleted_at}</Segment>
              </SegmentGroup>
              ))
            : null}
          </Segment.Group>
          <Container className="buttonsContainer">
            <Button floated="left" onClick={() => handleRestoreTeam(selectedTeam)}>
              Restore
            </Button>
            <Button floated="right" onClick={handleAddNewTeam}>
              Add new
            </Button>
          </Container>
        </Modal.Content>
      </Modal>
    )
};

export default connect(
  null,
  {
    addTeamToTribe,
    restoreTeamToTribe
  }
)(RestoreTeam);