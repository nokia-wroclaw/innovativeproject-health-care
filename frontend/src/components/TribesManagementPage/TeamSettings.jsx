import React, { useState } from "react";
import { connect } from "react-redux";
import { Modal, Button, Container, Input, Grid } from "semantic-ui-react";
import {
  deleteTeam,
  addManagerToTeam,
  deleteManagerFromTeam,
  addMemberToTeam,
  deleteMemberFromTeam,
  updateTeamName
} from "./../../store/actions/teams";
import { revalidateUser } from "../../store/actions/user";
import { confirmDialog } from "./../common/functions";
import EditingCard from "./../common/EditingCard/EditingCard";
import "../../styles/common.css";

const TeamSettings = ({ isOpen, team, close, ...props }) => {
  const [deleteBtnLoading, setDeleteBtnLoading] = useState(false);
  const [saveBtnLoading, setSaveBtnLoading] = useState(false);
  const [newTeamName, setNewTeamName] = useState(team.name);

  const handleDeleteTeam = () => {
    if (confirmDialog(team.name)) {
      setDeleteBtnLoading(true);
      props.deleteTeam(team);
    }
  };

  const handleAddManagerToTeam = user =>
    props.revalidateUser(user, props.addManagerToTeam(team, user));

  const handleDeleteManagerFromTeam = user =>
    props.revalidateUser(user, props.deleteManagerFromTeam(team, user));

  const handleAddMemberToTeam = user =>
    props.revalidateUser(user, props.addMemberToTeam(team, user));

  const handleDeleteMemberFromTeam = user =>
    props.revalidateUser(user, props.deleteMemberFromTeam(team, user));

  const handleSaveAndClose = async () => {
    setSaveBtnLoading(true);
    if (newTeamName !== team.name)
      await props.updateTeamName(team, newTeamName);
    setSaveBtnLoading(false);
    close();
  };

  return (
    <Modal open={isOpen}>
      <Modal.Header>
        <Grid stackable>
          <Grid.Column width={8}>
            <Input
              label={{ icon: "edit" }}
              labelPosition="right corner"
              defaultValue={team.name}
              onChange={({ target }) => setNewTeamName(target.value)}
            />
          </Grid.Column>
          <Grid.Column width={8}>
            <Button
              icon="trash alternate"
              labelPosition="left"
              floated="right"
              content="Delete Team"
              basic
              negative
              onClick={handleDeleteTeam}
              loading={deleteBtnLoading}
            />
          </Grid.Column>
        </Grid>
      </Modal.Header>

      <Modal.Content>
        <Container textAlign="left">
          <div className="flex-space-evenly-align-start">
            <EditingCard
              data={team.managers ? team.managers : []}
              title="Managers"
              useUsersForm={true}
              onAddBtnClick={handleAddManagerToTeam}
              onItemDelete={handleDeleteManagerFromTeam}
            />
            <EditingCard
              data={team.members ? team.members : []}
              title="Members"
              useUsersForm={true}
              onAddBtnClick={handleAddMemberToTeam}
              onItemDelete={handleDeleteMemberFromTeam}
            />
          </div>
        </Container>
        <Container textAlign="center">
          <Button onClick={handleSaveAndClose} primary loading={saveBtnLoading}>
            Save and close
          </Button>
        </Container>
      </Modal.Content>
    </Modal>
  );
};

const mapStateToProps = state => ({
  tribes: state.tribes
});

export default connect(
  mapStateToProps,
  {
    deleteTeam,
    addManagerToTeam,
    deleteManagerFromTeam,
    addMemberToTeam,
    deleteMemberFromTeam,
    updateTeamName,
    revalidateUser
  }
)(TeamSettings);
