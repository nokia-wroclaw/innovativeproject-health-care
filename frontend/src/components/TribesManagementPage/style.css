.tribes-accordion {
  border: 1px #aaa solid !important;
  background-color: rgba(0, 0, 0, 0.01) !important;
}

.teams-accordion {
  background-color: rgba(0, 0, 0, 0.02) !important;
  border: 1px #aaa solid !important;
}

.hovered {
  background: #f3f3f3f8 !important;
}

.clicked {
  background: #ecececf8 !important;
}

.buttonsContainer {
  padding-bottom: 30px;
}