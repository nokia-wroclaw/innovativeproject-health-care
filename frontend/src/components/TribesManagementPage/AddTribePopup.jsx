import React from "react";
import { Button, Icon, Popup } from "semantic-ui-react";
import "../../styles/common.css";
import AddTribeForm from "./AddTribeForm";

export const AddTribePopup = () => {
  return (
    <Popup
      trigger={
        <Button icon labelPosition="left" floated="right" compact secondary>
          <Icon name="plus" />
          New tribe
        </Button>
      }
      content={<AddTribeForm />}
      on="click"
      position="bottom center"
    />
  );
};

export default AddTribePopup;
