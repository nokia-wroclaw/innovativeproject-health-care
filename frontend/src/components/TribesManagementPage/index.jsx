import TribesManagementPage from './TribesManagementPage';
export default TribesManagementPage;
