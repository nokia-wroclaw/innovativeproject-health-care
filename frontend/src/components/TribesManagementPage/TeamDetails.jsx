import React, { useState } from "react";
import { connect } from "react-redux";
import { Button, Item, Label, List } from "semantic-ui-react";
import TeamSettings from "./TeamSettings";
import "../../styles/common.css";

const TeamDetails = ({ team, ...props }) => {
  const [isOpenSettings, setIsOpenSettings] = useState(false);

  return (
    <React.Fragment>
      <Button
        content="Team settings"
        icon="cog"
        labelPosition="left"
        compact
        secondary
        basic
        onClick={() => setIsOpenSettings(true)}
      />
      <TeamSettings
        isOpen={isOpenSettings}
        team={team}
        close={() => setIsOpenSettings(false)}
      />
      <Item style={{ paddingTop: "1em" }}>
        Managers ({team.managers ? team.managers.length : 0}):
        <br />
        {team.managers
          ? team.managers.map(manager => (
              <Label color="blue" key={manager.id}>
                {manager.name}
                <Label.Detail>({manager.login})</Label.Detail>
              </Label>
            ))
          : null}
      </Item>

      <List verticalAlign="middle">
        <List.Header>
          Members ({team.members ? team.members.length : 0}):
        </List.Header>

        {team.members
          ? team.members.map(member => (
              <List.Item key={member.id}>
                <List.Icon name="user circle" />
                {member.name} ({member.login})
              </List.Item>
            ))
          : null}
      </List>
    </React.Fragment>
  );
};

const mapStateToProps = state => ({
  tribes: state.tribes
});

export default connect(mapStateToProps)(TeamDetails);
