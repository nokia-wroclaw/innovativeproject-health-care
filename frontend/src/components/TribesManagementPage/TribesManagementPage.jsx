import React, { useState, useEffect } from "react";
import { connect } from "react-redux";
import { Accordion, Container, Header, Item } from "semantic-ui-react";
import { setTribes } from "./../../store/actions/tribes";
import TemplatePage from "../common/TemplatePage/";
import TribeDetails from "./TribeDetails";
import AddTribePopup from "./AddTribePopup";
import Loader from "./../common/Loader";

export const TribesManagementPage = props => {
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    props.setTribes().finally(() => setIsLoading(false));
  }, []);

  const tribePanels = [
    ...props.tribes.map(tribe => {
      const details = <TribeDetails tribe={tribe} />;
      return {
        key: tribe.id,
        title: tribe.name,
        content: { content: details }
      };
    })
  ];

  return (
    <TemplatePage>
      <Container>
        <Item style={{ margin: "1em 0" }}>
          {isLoading ? null : <AddTribePopup />}
          <Header as="h3">Your tribes</Header>
        </Item>
        {isLoading ? (
          <Loader />
        ) : (
          <Accordion
            className="tribes-accordion"
            fluid
            styled
            panels={tribePanels}
            exclusive={false}
          />
        )}
      </Container>
    </TemplatePage>
  );
};

const mapStateToProps = state => ({
  tribes: state.tribes
});

export default connect(
  mapStateToProps,
  { setTribes }
)(TribesManagementPage);
