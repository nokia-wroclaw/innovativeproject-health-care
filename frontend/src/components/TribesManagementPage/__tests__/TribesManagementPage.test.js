import React from "react";
import { shallow } from "enzyme";
import { TribesManagementPage } from "../TribesManagementPage";

const mockTribes = [{ id: 1, name: "tribe1" }, { id: 2, name: "tribe2" }];

describe("TribesManagementPage", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(<TribesManagementPage tribes={mockTribes} />);
    expect(wrapper).toMatchSnapshot();
  });
});
