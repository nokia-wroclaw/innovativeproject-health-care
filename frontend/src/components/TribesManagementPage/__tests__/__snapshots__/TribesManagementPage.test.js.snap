// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TribesManagementPage should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TribesManagementPage
    tribes={
      Array [
        Object {
          "id": 1,
          "name": "tribe1",
        },
        Object {
          "id": 2,
          "name": "tribe2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Container>
        <Item
          style={
            Object {
              "margin": "1em 0",
            }
          }
        >
          <Header
            as="h3"
          >
            Your tribes
          </Header>
        </Item>
        <Loader />
      </Container>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Item
            style={
              Object {
                "margin": "1em 0",
              }
            }
          >
            <Header
              as="h3"
            >
              Your tribes
            </Header>
          </Item>,
          <Loader />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              null,
              <Header
                as="h3"
              >
                Your tribes
              </Header>,
            ],
            "style": Object {
              "margin": "1em 0",
            },
          },
          "ref": null,
          "rendered": Array [
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "as": "h3",
                "children": "Your tribes",
              },
              "ref": null,
              "rendered": "Your tribes",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Container>
          <Item
            style={
              Object {
                "margin": "1em 0",
              }
            }
          >
            <Header
              as="h3"
            >
              Your tribes
            </Header>
          </Item>
          <Loader />
        </Container>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Item
              style={
                Object {
                  "margin": "1em 0",
                }
              }
            >
              <Header
                as="h3"
              >
                Your tribes
              </Header>
            </Item>,
            <Loader />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                null,
                <Header
                  as="h3"
                >
                  Your tribes
                </Header>,
              ],
              "style": Object {
                "margin": "1em 0",
              },
            },
            "ref": null,
            "rendered": Array [
              null,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": "h3",
                  "children": "Your tribes",
                },
                "ref": null,
                "rendered": "Your tribes",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
